File: docs/api-audit/probes/cpp/UniTensor_06_element.cpp
// Raw-C++ probe for the UniTensor cat-06 (element & block access)
// binding-fidelity findings. Links against a locally-built libcytnx so it can
// call the C++ methods DIRECTLY, bypassing the pybind layer — proving that:
//   * UT-E1: the C++ get_elem<T> template covers ALL 11 element dtypes (int,
//     uint, bool AND float/complex), so the Python get_elem lambda's 4-dtype
//     (float/complex-only) limit is a BINDING choice, not a C++ gap;
//   * UT-E2: the accessor methods get(accessors) / set(accessors, rhs) EXIST as
//     C++ methods (Python reaches them only via __getitem__/__setitem__, which
//     are the only public spelling — get/set are absent from dir(UniTensor));
//   * UT-E3: get_block_ returns a SHARED-DATA view (const Tensor& onto the
//     tensor's block) while get_block returns a COPY (clone).
#include "cytnx.hpp"
#include <iostream>
#include <vector>
#include <string>
using namespace cytnx;

static int fails = 0;
static void report(const std::string& claim, bool ok) {
  std::cout << (ok ? "[PASS] " : "[FAIL] ") << claim << std::endl;
  if (!ok) ++fails;
}

// A round-trip through the C++ get_elem<T>/set_elem<T> templates on a
// 1-element UniTensor of the matching storage dtype `dt`: writes 1, reads it
// back, and returns true iff the value survives. get_elem/set_elem are
// [[deprecated]] in C++ ("Use at() instead."), so the caller silences that.
template <class T>
static bool roundtrip_dtype(int dt) {
  UniTensor U(zeros({1, 1}, dt));
  U.set_elem<T>({0, 0}, (T)1);
  return U.get_elem<T>({0, 0}) == (T)1;
}

int main() {
  // -----------------------------------------------------------------------
  // UT-E1: C++ `get_elem<T>` (and `set_elem<T2>`) are generic TEMPLATES (they
  // forward to at<T>()), instantiable for every element dtype T. Round-trip a
  // value through all 11 storage dtypes — including the 7 integer/bool dtypes
  // the Python get_elem lambda REJECTS (it only wires up the 4 float/complex
  // branches). So the Python 4-dtype limit is a BINDING choice, not a C++ gap.
  // -----------------------------------------------------------------------
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
  {
    bool all_ok =
      roundtrip_dtype<cytnx_double>(Type.Double) &&
      roundtrip_dtype<cytnx_float>(Type.Float) &&
      roundtrip_dtype<cytnx_complex128>(Type.ComplexDouble) &&
      roundtrip_dtype<cytnx_complex64>(Type.ComplexFloat) &&
      roundtrip_dtype<cytnx_int64>(Type.Int64) &&
      roundtrip_dtype<cytnx_uint64>(Type.Uint64) &&
      roundtrip_dtype<cytnx_int32>(Type.Int32) &&
      roundtrip_dtype<cytnx_uint32>(Type.Uint32) &&
      roundtrip_dtype<cytnx_int16>(Type.Int16) &&
      roundtrip_dtype<cytnx_uint16>(Type.Uint16) &&
      roundtrip_dtype<cytnx_bool>(Type.Bool);
    report("UT-E1: C++ get_elem<T>/set_elem<T> are templates covering ALL 11 "
           "element dtypes (a value round-trips on Double/Float/ComplexDouble/"
           "ComplexFloat AND on all 7 integer/bool dtypes) -- the Python "
           "get_elem's 4-dtype (float/complex-only) limit is a binding choice, "
           "not a C++ gap",
           all_ok);
  }

  // The asymmetry is ONLY on the get_elem read side: the Python set_elem/item
  // bindings already cover all 11 (as their C++ templates do here).
  {
    bool ok_b = roundtrip_dtype<cytnx_bool>(Type.Bool);
    UniTensor V(zeros({1, 1}, Type.Double));
    V.set_elem<cytnx_double>({0, 0}, 2.5);
    bool ok_d = V.get_elem<cytnx_double>({0, 0}) == 2.5;
    report("the C++ get/set element templates handle both a bool tensor and a "
           "fractional double cleanly -- confirming the dtype asymmetry lives "
           "only in the Python get_elem binding, not in C++ (UT-E1)",
           ok_b && ok_d);
  }
#pragma GCC diagnostic pop

  // -----------------------------------------------------------------------
  // UT-E2: the accessor methods `get(accessors)` / `set(accessors, rhs)` EXIST
  // as C++ methods. Python does NOT bind them by name (they are absent from
  // dir(UniTensor)); the only public Python path is __getitem__/__setitem__,
  // whose lambdas call exactly these. Here we call them DIRECTLY.
  // -----------------------------------------------------------------------
  {
    // get(): slice row 0 of a [2,3] Dense tensor -> shape [3].
    UniTensor U(arange(6).reshape(2, 3));
    UniTensor sub = U.get({Accessor(0), Accessor::all()});
    report("UT-E2: C++ get(accessors) EXISTS and slices a Dense tensor "
           "(U.get({0, :}) -> shape [3]) -- Python reaches it only via "
           "__getitem__ (get is not a public Python member)",
           sub.shape() == std::vector<cytnx_uint64>{3});
  }
  {
    // set(): assign a row of a [2,3] Dense tensor from a [3] Tensor.
    UniTensor U(zeros({2, 3}));
    Tensor rhs = ones({3});
    UniTensor& r = U.set({Accessor(0), Accessor::all()}, rhs);
    bool wrote = double(U.at<double>({0, 0}) == 1.0) &&
                 double(U.at<double>({1, 0}) == 0.0);
    report("UT-E2: C++ set(accessors, rhs) EXISTS and assigns into a Dense "
           "tensor (row 0 becomes 1, row 1 stays 0) returning *this -- Python "
           "reaches it only via __setitem__ (set is not a public Python member)",
           (&r == &U) && wrote);
  }

  // -----------------------------------------------------------------------
  // UT-E3: get_block_ returns a SHARED-DATA view; get_block returns a COPY.
  // -----------------------------------------------------------------------
  {
    UniTensor U(ones({2, 2}));
    Tensor& view = U.get_block_();   // reference into the tensor's block
    Tensor copy = U.get_block();     // a clone
    report("UT-E3: C++ get_block_ returns a shared-data VIEW (same_data with the "
           "tensor's block) while get_block returns an independent COPY "
           "(not same_data)",
           U.get_block_().same_data(view) && !U.get_block_().same_data(copy));
  }

  std::cout << (fails ? "cpp probe FAILED" : "cpp probe ok") << std::endl;
  return fails ? 1 : 0;
}

File: docs/api-audit/probes/UniTensor_06_element.py
"""Behavioral probe for UniTensor category 06 — element & block access,
verified against the installed cytnx==1.1.0 wheel (NOT source-inferred).

Every runtime claim in docs/api-audit/UniTensor/06-element-block-access.md is
backed by a report(...) assertion here. The raw-C++ side of the
binding-fidelity findings is verified by probes/cpp/UniTensor_06_element.cpp.
Run: source tools/env.sh && $PY docs/api-audit/probes/UniTensor_06_element.py
"""
import sys, os, warnings
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "tools"))
import cytnx
from probe_helper import report

UT = cytnx.UniTensor

# The 11 element dtypes cytnx supports.
ALL_DTYPES = [
    cytnx.Type.ComplexDouble, cytnx.Type.ComplexFloat, cytnx.Type.Double,
    cytnx.Type.Float, cytnx.Type.Int64, cytnx.Type.Uint64, cytnx.Type.Int32,
    cytnx.Type.Uint32, cytnx.Type.Int16, cytnx.Type.Uint16, cytnx.Type.Bool,
]
FLOAT_COMPLEX = {cytnx.Type.ComplexDouble, cytnx.Type.ComplexFloat,
                 cytnx.Type.Double, cytnx.Type.Float}


def mk_scalar(dtype):
    """A fresh 1-element Dense UniTensor of the given dtype (for item/get/set)."""
    return UT.zeros([1, 1], dtype=dtype)


def mk_dense():
    """A fresh rank-2 Dense (bosonic) UniTensor filled 1."""
    return UT.ones([2, 2])


def mk_block():
    """A fresh rank-2 Block (U1-symmetric) UniTensor with two 1x1 blocks."""
    sym = cytnx.Symmetry.U1()
    Bi = cytnx.Bond(cytnx.BD_IN, [[0], [1]], [1, 1], [sym])
    Bo = cytnx.Bond(cytnx.BD_OUT, [[0], [1]], [1, 1], [sym])
    return cytnx.UniTensor([Bi, Bo], labels=["a", "b"])


report("wheel under test is 1.1.0 (runtime, not source-inferred)",
       cytnx.__version__ == "1.1.0")

# =========================================================================
# UT-E1 — get_elem binds only the 4 float/complex dtypes, while item/set_elem
#          cover all 11 (binding fidelity: the pybind template coverage differs)
# =========================================================================

# get_elem works on the 4 float/complex dtypes.
for dt in FLOAT_COMPLEX:
    u = mk_scalar(dt)
    v = u.get_elem([0, 0])
    report(f"get_elem returns a value on float/complex dtype {dt.name}",
           v == 0)

# get_elem RAISES on the 7 integer/bool dtypes (the pybind lambda has no branch
# for them and falls through to `[ERROR] try to get element from a void Storage`).
int_bool = [dt for dt in ALL_DTYPES if dt not in FLOAT_COMPLEX]
for dt in int_bool:
    u = mk_scalar(dt)
    raised = False
    try:
        u.get_elem([0, 0])
    except RuntimeError:
        raised = True
    report(f"get_elem RAISES on integer/bool dtype {dt.name} — "
           f"the binding lambda only instantiates the 4 float/complex branches",
           raised)

# item() and set_elem() cover ALL 11 dtypes — including the integer/bool ones
# get_elem rejects. This is the binding-fidelity asymmetry (C++ get_elem<T> is a
# generic template covering all 11; the Python get_elem lambda picks only 4).
for dt in ALL_DTYPES:
    u = mk_scalar(dt)
    report(f"item() reads a value on dtype {dt.name} (all 11 covered)",
           u.item() == 0)
    u.set_elem([0, 0], 1)
    report(f"set_elem writes a value on dtype {dt.name} (all 11 covered)",
           u.item() == 1)

# =========================================================================
# UT-E2 — the C++ `get`/`set` accessor methods are UNBOUND in Python; they are
#          reached only through __getitem__/__setitem__ (binding fidelity)
# =========================================================================

report("the raw C++ accessor methods `get`/`set` are NOT public Python members "
       "— they are reachable only via __getitem__/__setitem__",
       "get" not in dir(UT) and "set" not in dir(UT))
report("__getitem__ and __setitem__ ARE bound (the only public path to C++ get/set)",
       "__getitem__" in dir(UT) and "__setitem__" in dir(UT))

# =========================================================================
# UT-E3 — get_block (copy) vs get_block_ (shared-data view) on a Dense tensor
# =========================================================================

u = mk_dense()
ref = u.get_block_()          # a view onto the tensor's storage
copy = u.get_block()          # a clone
report("get_block returns a COPY: its data is NOT shared with the tensor's "
       "block storage (same_data is False)",
       not ref.same_data(copy))
report("get_block_ returns a shared-data VIEW: its data IS shared with the "
       "tensor's block storage (same_data is True)",
       ref.same_data(u.get_block_()))

# =========================================================================
# UT-E4 — get_blocks (copy) vs get_blocks_ (shared-data view) on a Block tensor.
#          get_blocks(_) ERROR on a Dense tensor (use get_block(_) instead).
# =========================================================================

bt = mk_block()
ref_blocks = bt.get_blocks_()
copies = bt.get_blocks()
views = bt.get_blocks_()
report("get_blocks returns COPIES: none of the returned tensors share data with "
       "the tensor's block storage",
       len(copies) == len(ref_blocks) and
       not any(ref_blocks[i].same_data(copies[i]) for i in range(len(copies))))
report("get_blocks_ returns shared-data VIEWS: every returned tensor shares data "
       "with the tensor's block storage",
       len(views) == len(ref_blocks) and
       all(ref_blocks[i].same_data(views[i]) for i in range(len(views))))

dense_err = False
try:
    mk_dense().get_blocks()
except RuntimeError:
    dense_err = True
report("get_blocks errors on a Dense tensor ('use get_block() instead') — the "
       "plural forms are for Block/BlockFermionic tensors only",
       dense_err)

# =========================================================================
# UT-E5 — the misspelled `slient` kwarg on get_blocks_ warns (FutureWarning) and
#          still forwards; the correctly-spelled `silent` does not warn
# =========================================================================

with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    r = mk_block().get_blocks_(slient=True)
report("the misspelled `slient` kwarg on get_blocks_ emits a FutureWarning "
       "('deprecated ... use silent instead') yet still forwards the call",
       any(issubclass(x.category, FutureWarning) and "slient" in str(x.message)
           for x in w))

with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    mk_block().get_blocks_(silent=True)
report("the correctly-spelled `silent` kwarg on get_blocks_ emits NO warning",
       len(w) == 0)

# =========================================================================
# UT-E6 — put_block(..., force=...) / put_block_(..., force=...) are DEPRECATED
#          (FutureWarning); the force-free overloads are the current form
# =========================================================================

bt = mk_block()
blk = bt.get_block([0, 0])
with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    bt.put_block(blk, [0, 0], True)          # positional force -> deprecated overload
report("put_block(in, qidx, force) emits a FutureWarning — the `force` argument "
       "is deprecated ('use put_block(in, qnum) without force instead')",
       any(issubclass(x.category, FutureWarning) and "force" in str(x.message)
           for x in w))

bt = mk_block()
blk = bt.get_block([0, 0])
with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    bt.put_block_(blk, [0, 0], True)
report("put_block_(in, qidx, force) likewise emits a FutureWarning for the "
       "deprecated `force` argument",
       any(issubclass(x.category, FutureWarning) and "force" in str(x.message)
           for x in w))

# =========================================================================
# UT-E7 — put_block (copy-in) vs put_block_ (view-in): a correct N-underscore
#          pair. Both return None (they mutate the receiver in place).
# =========================================================================

u = mk_dense()
t = cytnx.ones([2, 2])
ret = u.put_block(t)
report("put_block COPIES the input tensor into the block (the receiver's block "
       "does NOT share data with the input) and returns None",
       ret is None and not u.get_block_().same_data(t))

u = mk_dense()
t = cytnx.ones([2, 2])
ret = u.put_block_(t)
report("put_block_ makes the input tensor a shared-data VIEW of the block (the "
       "receiver's block SHARES data with the input) and returns None",
       ret is None and u.get_block_().same_data(t))

# =========================================================================
# UT-E8 — __getitem__/__setitem__ slice a Dense tensor; they ERROR on a
#          Block/BlockFermionic tensor (use at()/get_block() instead)
# =========================================================================

u = UT.arange(6).reshape(2, 3)
sl = u[0, :]
report("__getitem__ slices a Dense tensor, returning a UniTensor sub-block "
       "(u[0, :] has shape [3])",
       isinstance(sl, UT) and sl.shape() == [3])

u = UT.arange(6).reshape(2, 3)
u[0, 0] = 9.0
report("__setitem__ assigns into a Dense tensor (u[0,0] = 9.0 sets the element)",
       u.get_elem([0, 0]) == 9.0)

block_getitem_err = False
try:
    mk_block()[0, 0]
except RuntimeError:
    block_getitem_err = True
report("__getitem__ ERRORS on a Block tensor ('Cannot get element using [] from "
       "Block/BlockFermionicUniTensor. Use at() instead.')",
       block_getitem_err)

# =========================================================================
# UT-E9 — at()/item() element access. at() returns a proxy usable for read AND
#          write; it works on BOTH Dense and Block tensors (unlike []).
# =========================================================================

u = mk_dense()
h = u.at([0, 0])
report("at([0,0]) on a Dense tensor returns a proxy whose .value reads the "
       "element",
       h.value == 1.0)
h.value = 3.0
report("the at() proxy is writable: setting proxy.value mutates the element in "
       "place",
       u.get_elem([0, 0]) == 3.0)

bt = mk_block()
hb = bt.at([0, 0])
report("at([0,0]) works on a Block tensor too (where [] is rejected): the proxy "
       "reports whether the element exists",
       hb.exists())

# =========================================================================
# UT-E10 — elem_exists is a Block-only predicate: it reports whether a symmetry
#           block-element exists, and ERRORS on a Dense tensor
# =========================================================================

bt = mk_block()
report("elem_exists reports True for an allowed block-element [0,0] and False "
       "for a symmetry-forbidden one [0,1]",
       bt.elem_exists([0, 0]) and not bt.elem_exists([0, 1]))

dense_ee_err = False
try:
    mk_dense().elem_exists([0, 0])
except RuntimeError:
    dense_ee_err = True
report("elem_exists ERRORS on a Dense tensor ('can only be used on UniTensor "
       "with Symmetry') — it is a Block-only predicate",
       dense_ee_err)

print("UniTensor 06 probe ok")
